{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"mount_file_id":"1_9i5qxObBP60JSeCfvd-3wCrFbaoP2Dm","authorship_tag":"ABX9TyP4PXmuIsBbvs9e3tEI+j9h"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["# import packages\n","try:\n","  from Bio import SeqIO\n","except ImportError as e:\n","  !pip install biopython\n","  from Bio import SeqIO\n","import pickle\n","import os"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"h2dIUXl9I6Gr","executionInfo":{"status":"ok","timestamp":1683435929871,"user_tz":420,"elapsed":8449,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"}},"outputId":"5e6a0d8d-e6ba-4d03-8630-0576f17406fb"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting biopython\n","  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m36.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.22.4)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.81\n"]}]},{"cell_type":"code","execution_count":null,"metadata":{"id":"fNbgHzUnIomj"},"outputs":[],"source":["# configs\n","gene_length_max = 5000 # temporary max limit on gene sequence length\n","feature_types = ('CDS', 'mat_peptide')\n","downloaded_strains = [\n","  'NC_045512.2',\n","  'OP733821.1',\n","  'OK341237.1',\n","  'OM251163.1',\n","  'OQ050563.1',\n","  'MW474188.1',\n","  'MW243586.1',\n","  'OL947440.1',\n","  'OQ253610.1'\n","]"]},{"cell_type":"code","source":["# the ncbi page text for each strain was downloaded along with their fasta file\n","# use the files to gather data for each strain and store it to a .pickle file\n","for strain in downloaded_strains:\n","\n","  # variables that change for each strain\n","  ncbi_page_path = './drive/MyDrive/Colab Notebooks/ncbi_pages/' + strain + '.txt'\n","  fasta_path = './drive/MyDrive/Colab Notebooks/fasta_files/' + strain + '.fasta'\n","  pickles_path = './drive/MyDrive/Colab Notebooks/pickles/'\n","  pickles_file = strain + '.pickle'\n","\n","  # find lines of text with info about a feature\n","  # put list of lines for each feature into a list\n","  features = []\n","  with open(ncbi_page_path) as file:\n","    found = False\n","    feature_lines = []\n","    for line in file:\n","      line = line.strip()\n","      if found == True and line.startswith('/'):\n","        if line.startswith('/translation'):\n","          continue\n","        feature_lines.append(line)\n","        continue\n","      elif found == True:\n","        features.append(feature_lines)\n","        feature_lines = []\n","        found = False    \n","      for ftype in feature_types:\n","        if line.startswith(ftype):\n","          found = True    \n","          feature_lines.append(line)\n","\n","  # go through gathered text lines about features\n","  # put useful data into a dict structure and put those in a list\n","  dna_features = []\n","  feature = {}\n","  for feature_lines in features:\n","    top_line = \" \".join(feature_lines[0].split())\n","    line_parts = top_line.split(' ')\n","    if len(line_parts) != 2:\n","      continue\n","    feature_type, cords_text = line_parts\n","    # TODO: handle join cases: mat_peptide join(13442..13468,13468..16236)\n","    coords = cords_text.split('..')\n","    if len(coords) != 2:\n","      continue\n","    start, end = coords\n","    feature['type'] = feature_type\n","    # TODO: handle greater than cases: '>27123'\n","    if not start.isdigit() or not end.isdigit():\n","      continue\n","    feature['start'] = int(start)\n","    feature['end'] = int(end)\n","    for line in feature_lines[1:]:\n","      key, val = line.split('=')\n","      key = key[1:]\n","      val = val.strip('\",')\n","      feature[key] = val\n","    dna_features.append(feature)\n","    feature = {}\n","  \n","  # use start and end porperties found in text lines\n","  # to find protein coding sequences from fasta file\n","  fasta = SeqIO.parse(fasta_path,\"fasta\")\n","  records = []\n","  for record in fasta.records:\n","    records.append(record)\n","  dna = records[0].seq\n","  for feature in dna_features:\n","    start = feature['start'] - 1\n","    end = feature['end']\n","    if end - start < gene_length_max:\n","      coding_region = dna[start:end]\n","      feature['nucleotides'] = coding_region\n","      feature['translation'] = coding_region.translate()\n","\n","  # store the list of found features in a .pickle file\n","  with open(pickles_path + pickles_file, 'wb') as file:\n","    pickle.dump(dna_features, file)"],"metadata":{"id":"EZPaEDkBI1md","executionInfo":{"status":"error","timestamp":1683867544133,"user_tz":420,"elapsed":890,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"}},"colab":{"base_uri":"https://localhost:8080/","height":240},"outputId":"ed99f02f-a7bd-4d2f-cde7-47aed7c9a83f"},"execution_count":1,"outputs":[{"output_type":"error","ename":"NameError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-1-d549b6c683ad>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# the ncbi page text for each strain was downloaded along with their fasta file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m# use the files to gather data for each strain and store it to a .pickle file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mstrain\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdownloaded_strains\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m   \u001b[0;31m# variables that change for each strain\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mNameError\u001b[0m: name 'downloaded_strains' is not defined"]}]},{"cell_type":"code","source":["os.listdir(pickles_path)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"V30Y26zrJyeC","executionInfo":{"status":"ok","timestamp":1683435966325,"user_tz":420,"elapsed":343,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"}},"outputId":"5fb7b254-daed-479e-bfdc-35080aba9115"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['OP733821.1.pickle',\n"," 'OL947440.1.pickle',\n"," 'OK341237.1.pickle',\n"," 'MW474188.1.pickle',\n"," 'MW243586.1.pickle',\n"," 'OQ050563.1.pickle',\n"," 'OQ253610.1.pickle',\n"," '.ipynb_checkpoints',\n"," 'NC_045512.2.pickle',\n"," 'OM251163.1.pickle']"]},"metadata":{},"execution_count":4}]},{"cell_type":"code","source":[],"metadata":{"id":"dPNYJRDvsNWP"},"execution_count":null,"outputs":[]}]}