{"cells":[{"attachments":{},"cell_type":"markdown","metadata":{},"source":["# Process NCBI Data Files\n","\n","The purpose of this notebook is to gather the nucleotide sequences for COVID-19 features (e.g. proteins). Feature coordinates are read from text files manually downloaded from NCBI (folder `ncbi_pages`). FASTA files (folder `fasta_files`) for each strain are then repeatedly referenced to tranlate specific nucleotide ranges to amino acid sequences for future analysis.\n","\n","Notebook workflow:\n","1. Scan each strain's NCBI text file for the start-end coordinates of each listed \"feature\" type \n","2. Use these coordinates to find the nucleotide sequences and translate them to amino acids\n","3. Store sequence results as structured data in \".pickle\" files located in the `pickles` folder\n","4. Pickle files are used to save Python objects to files that can be stored on disk and used as inputs to other Python scripts."]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Imports\n","\n","Begin workflow by importing the open source Python libraries to be used."]},{"cell_type":"code","execution_count":9,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":8449,"status":"ok","timestamp":1683435929871,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"},"user_tz":420},"id":"h2dIUXl9I6Gr","outputId":"5e6a0d8d-e6ba-4d03-8630-0576f17406fb"},"outputs":[{"name":"stdout","output_type":"stream","text":["Library imports complete\n"]}],"source":["try:\n","  from Bio import SeqIO\n","except ImportError as e:\n","  %pip install biopython\n","  from Bio import SeqIO\n","import pickle\n","import os\n","\n","print('Library imports complete')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Configs\n","\n","Set the config variables as needed."]},{"cell_type":"code","execution_count":10,"metadata":{"id":"fNbgHzUnIomj"},"outputs":[{"name":"stdout","output_type":"stream","text":["Config variables set\n"]}],"source":["# folder relative locations\n","ncbi_page_folder = '../ncbi_pages/'\n","fasta_folder = '../fasta_files/'\n","pickles_folder = '../pickles/'\n","\n","# temporary max limit on gene sequence length to avoid long computations\n","gene_length_max = 5000 \n","\n","# feature type labels to search for in NCBI text pages\n","feature_types = ('CDS', 'mat_peptide')\n","\n","# list of strain ascession numbers matching downloaded ncbi and fasta files\n","downloaded_strains = [\n","  'NC_045512.2',\n","  'OP733821.1',\n","  'OK341237.1',\n","  'OM251163.1',\n","  'OQ050563.1',\n","  'MW474188.1',\n","  'MW243586.1',\n","  'OL947440.1',\n","  'OQ253610.1'\n","]\n","\n","print('Config variables set')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Process\n","\n","The NCBI page text for each strain was previously downloaded along with their FASTA file. \n","1. Open each NCBI page text file and gather the lines about DNA features\n","2. Scan through those lines to find the feature names and their start-end coordinates\n","3. Open the strain's FASTA file and find the sequences using the start-end coordinates\n","4. Use biopython to translate the nucleotide sequences to amino acids\n","5. Save the gathered data for each strain to a separate .pickle file for each strain"]},{"cell_type":"code","execution_count":13,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":240},"executionInfo":{"elapsed":890,"status":"error","timestamp":1683867544133,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"},"user_tz":420},"id":"EZPaEDkBI1md","outputId":"ed99f02f-a7bd-4d2f-cde7-47aed7c9a83f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Features found for NC_045512.2 strain: 36\n","Features found for OP733821.1 strain: 36\n","Features found for OK341237.1 strain: 36\n","Features found for OM251163.1 strain: 35\n","Features found for OQ050563.1 strain: 36\n","Features found for MW474188.1 strain: 36\n","Features found for MW243586.1 strain: 36\n","Features found for OL947440.1 strain: 36\n","Features found for OQ253610.1 strain: 35\n","\n","Processing and saving data complete\n"]}],"source":["\n","for strain in downloaded_strains:\n","\n","  # variables that change for each strain\n","  ncbi_page_path = ncbi_page_folder + strain + '.txt'\n","  fasta_path = fasta_folder + strain + '.fasta'\n","  pickles_file = strain + '.pickle'\n","\n","  # find lines of text with info about a feature\n","  # put list of lines for each feature into a list\n","  features = []\n","  with open(ncbi_page_path) as file:\n","    found = False\n","    feature_lines = []\n","    for line in file:\n","      line = line.strip()\n","      if found == True and line.startswith('/'):\n","        if line.startswith('/translation'):\n","          continue\n","        feature_lines.append(line)\n","        continue\n","      elif found == True:\n","        features.append(feature_lines)\n","        feature_lines = []\n","        found = False    \n","      for ftype in feature_types:\n","        if line.startswith(ftype):\n","          found = True    \n","          feature_lines.append(line)\n","\n","  # go through gathered text lines about features\n","  # put useful data into a dict structure and put those in a list\n","  dna_features = []\n","  feature = {}\n","  for feature_lines in features:\n","    top_line = \" \".join(feature_lines[0].split())\n","    line_parts = top_line.split(' ')\n","    if len(line_parts) != 2:\n","      continue\n","    feature_type, cords_text = line_parts\n","    # TODO: handle join cases: mat_peptide join(13442..13468,13468..16236)\n","    coords = cords_text.split('..')\n","    if len(coords) != 2:\n","      continue\n","    start, end = coords\n","    feature['type'] = feature_type\n","    # TODO: handle greater than cases: '>27123'\n","    if not start.isdigit() or not end.isdigit():\n","      continue\n","    feature['start'] = int(start)\n","    feature['end'] = int(end)\n","    for line in feature_lines[1:]:\n","      key, val = line.split('=')\n","      key = key[1:]\n","      val = val.strip('\",')\n","      feature[key] = val\n","    dna_features.append(feature)\n","    feature = {}\n","  \n","  # use start and end porperties found in text lines\n","  # to find protein coding sequences from fasta file\n","  fasta = SeqIO.parse(fasta_path,\"fasta\")\n","  records = []\n","  for record in fasta.records:\n","    records.append(record)\n","  dna = records[0].seq\n","  for feature in dna_features:\n","    start = feature['start'] - 1\n","    end = feature['end']\n","    if end - start < gene_length_max:\n","      coding_region = dna[start:end]\n","      feature['nucleotides'] = coding_region\n","      feature['translation'] = coding_region.translate()\n","\n","  # store the list of found features in a .pickle file\n","  with open(pickles_folder + pickles_file, 'wb') as file:\n","    pickle.dump(dna_features, file)\n","\n","  # print the count of features found for strain\n","  print('Features found for {} strain: {}'.format(strain, len(dna_features)))\n","\n","print('\\nProcessing and saving data complete')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Results\n","\n","Each strain's features are saved as a list object and stored in .pickle file:"]},{"cell_type":"code","execution_count":14,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["MW243586.1.pickle\n","OP733821.1.pickle\n","OK341237.1.pickle\n","OM251163.1.pickle\n","OQ050563.1.pickle\n","MW474188.1.pickle\n","OL947440.1.pickle\n","NC_045512.2.pickle\n","OQ253610.1.pickle\n"]}],"source":["\n","# Print the names of features found\n","for file in os.listdir(pickles_folder):\n","  if ('.pickle' in file):\n","    print(file)"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyP4PXmuIsBbvs9e3tEI+j9h","mount_file_id":"1_9i5qxObBP60JSeCfvd-3wCrFbaoP2Dm","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.1"}},"nbformat":4,"nbformat_minor":0}
