{"cells":[{"attachments":{},"cell_type":"markdown","metadata":{},"source":["# Find COVID-19 Mutations\n","\n","The purpose of this notebook is find mutations in COVID-19 strains. This analyze depends on strain data compiled by the `process_strains.ipynb` notebook and stored in the `pickles` folder. \n","\n","Notebook workflow:\n","1. Open pickle folder and get sequences of the oringal strain to be compared to \n","2. Open pickle folder and get sequences for all other strains\n","3. Comapre the amino acid sequence of each strain to the orignal and save found differences as mutations\n","4. Re-save the picle files along with an new field for each feature called `mutants` for mutations"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Imports\n","\n","Begin workflow by importing the open source Python libraries to be used."]},{"cell_type":"code","execution_count":19,"metadata":{"executionInfo":{"elapsed":416,"status":"ok","timestamp":1684159131938,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"},"user_tz":420},"id":"OcFN-W9DyO2F"},"outputs":[{"name":"stdout","output_type":"stream","text":["Library imports complete\n"]}],"source":["try:\n","  from Bio import pairwise2\n","except ImportError as e:\n","  !pip install biopython\n","  from Bio import pairwise2\n","import pickle\n","import os\n","import pprint\n","pp = pprint.PrettyPrinter(indent=4)\n","from Bio.pairwise2 import MAX_ALIGNMENTS\n","\n","print('Library imports complete')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Configs\n","\n","Set the config variables as needed."]},{"cell_type":"code","execution_count":20,"metadata":{"id":"llq055B_Rtfq"},"outputs":[{"name":"stdout","output_type":"stream","text":["Config variables set\n"]}],"source":["original_strain = 'NC_045512.2'\n","\n","pickles_path = '../pickles/'\n","\n","amino_acid_colors = {\n","  'T': 'green',\n","  'Q': 'green',\n","  'S': 'green',\n","  'N': 'green',\n","  'F': 'blue',\n","  'M': 'blue',\n","  'L': 'blue',\n","  'V': 'blue',\n","  'W': 'blue',\n","  'A': 'blue',\n","  'I': 'blue',\n","  'E': 'purple',\n","  'D': 'purple',\n","  'K': 'red',\n","  'R': 'red',\n","  'C': 'peach',\n","  'G': 'peach',\n","  'P': 'yellow',\n","  'H': 'teal',\n","  'Y': 'teal'\n"," }  \n","amino_acid_types = {\n","  'T': 'polar',\n","  'Q': 'polar',\n","  'S': 'polar',\n","  'N': 'polar',\n","  'F': 'nonpolar',\n","  'M': 'nonpolar',\n","  'L': 'nonpolar',\n","  'V': 'nonpolar',\n","  'W': 'nonpolar',\n","  'A': 'nonpolar',\n","  'I': 'nonpolar',\n","  'E': 'negative',\n","  'D': 'negative',\n","  'K': 'positive',\n","  'R': 'positive',\n","  'C': 'nonpolar',\n","  'G': 'nonpolar',\n","  'P': 'nonpolar',\n","  'H': 'positive',\n","  'Y': 'polar'\n"," }\n","\n","print('Config variables set')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Original strain features\n","\n","Here we retrieve the found features that were stored previously in pickle files by the `process_strains.ipynb` notebook. In order to compare a feature accross strains, we need to make use of the feature `product` name. Each feature from NCBI has product name and nucleotide start-end coordinates. We can restructure the feature info into a dictionary using product name for lookup keys."]},{"cell_type":"code","execution_count":21,"metadata":{"id":"xJnEwAJkzQSg"},"outputs":[{"name":"stdout","output_type":"stream","text":["Original strain NC_045512.2 features read:\n","   leader protein\n","   nsp2\n","   nsp3\n","   nsp4\n","   3C-like proteinase\n","   nsp6\n","   nsp7\n","   nsp8\n","   nsp9\n","   nsp10\n","   helicase\n","   3'-to-5' exonuclease\n","   endoRNAse\n","   2'-O-ribose methyltransferase\n","   ORF1a polyprotein\n","   nsp11\n","   surface glycoprotein\n","   ORF3a protein\n","   envelope protein\n","   membrane glycoprotein\n","   ORF6 protein\n","   ORF7a protein\n","   ORF7b\n","   ORF8 protein\n","   nucleocapsid phosphoprotein\n","   ORF10 protein\n"]}],"source":["# Open original strain pickle file of feature data generated by\n","strain = original_strain\n","pickles_file = strain + '.pickle'\n","filepath = pickles_path + pickles_file\n","objects = []\n","with (open(filepath, \"rb\")) as openfile:\n","  while True:\n","    try:\n","      objects.append(pickle.load(openfile))\n","    except EOFError:\n","      break\n","orig_feats = objects[0]\n","\n","# Scan the info about each feature and store it to a dictionary named `original` indexed by its `product` name\n","original = {}\n","for feat in orig_feats:\n","  product = feat['product']\n","  if product not in original:\n","    original[product] = feat\n","\n","# Print the names of features found\n","print('Original strain {} features read:'.format(original_strain))\n","for product in original.keys():\n","  print('  ', product)"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### Features for the other strains\n","\n","Now we want to do the same for other strains. These are kept in their own list. The original is separate because the data each of the strains in this new list will be compared to the same original strain. Deviations from the amino acid sequence of the original strain will be considered mutations."]},{"cell_type":"code","execution_count":22,"metadata":{"id":"OjSG_QD1zW0D"},"outputs":[{"name":"stdout","output_type":"stream","text":["Mutations found for MW243586.1 strain: 23\n","Mutations found for OP733821.1 strain: 57\n","Mutations found for OK341237.1 strain: 7\n","Mutations found for OM251163.1 strain: 73\n","Mutations found for OQ050563.1 strain: 6\n","Mutations found for MW474188.1 strain: 1\n","Mutations found for OL947440.1 strain: 131\n","Mutations found for OQ253610.1 strain: 66\n","\n","Finding mutations and re-saving files complete\n"]}],"source":["for file_name in os.listdir(pickles_path):\n","  if '.pickle' not in file_name:\n","    continue\n","  comparison_strain = file_name.replace('.pickle', '')\n","  if comparison_strain == original_strain:\n","    continue\n","  strain = comparison_strain\n","  pickles_file = strain + '.pickle'\n","  filepath = pickles_path + pickles_file\n","  objects = []\n","  with (open(filepath, \"rb\")) as openfile:\n","    while True:\n","      try:\n","        objects.append(pickle.load(openfile))\n","      except EOFError:\n","        break\n","  comp_feats = objects[0]\n","\n","  comparison = {}\n","  for feat in comp_feats:\n","    product = feat['product']\n","    # rewrite in case of repeats should be same\n","    if product not in comparison:\n","      comparison[product] = feat\n","\n","  mutations_count = 0\n","  for key, comp_feat in comparison.items():\n","    orig_feat = original[key]\n","    if 'translation' not in comp_feat:\n","      continue\n","    alignments = pairwise2.align.globalxx(orig_feat['translation'], comp_feat['translation'])\n","    max_alignment = None\n","    max_score = 0\n","    for alignment in alignments:\n","      if alignment.score > max_score:\n","        max_score = alignment.score\n","        max_alignment = alignment\n","  \n","    str1 = max_alignment.seqA\n","    str2 = max_alignment.seqB\n","    positions = [] \n","    for x in range (0, min(len(str1), len(str2))): \n","      if str1[x].upper() != str2[x].upper(): \n","        positions.append(x)\n","\n","    diffs = []\n","    for position in positions:\n","      coord = position + 1\n","      diffs.append( (coord, str1[position], str2[position]) )\n","    comp_feat['diffs'] = diffs\n","\n","  for comp_feat in comparison.values():\n","    diffs = []\n","    if 'diffs' in comp_feat:\n","      diffs = comp_feat['diffs']\n","    else:\n","      comp_feat['diffs'] = diffs\n","    mutants = []\n","    for idx in range(len(diffs) - 1):\n","      left = diffs[idx][0]\n","      right = diffs[idx + 1][0]\n","      if diffs[idx][1] != '-':\n","        orig_amino_acid = diffs[idx][1]\n","      else:\n","        orig_amino_acid = diffs[idx][2]\n","      if diffs[idx + 1][2] != '-':\n","        comp_amino_acid = diffs[idx + 1][2]\n","      else:\n","        comp_amino_acid = diffs[idx + 1][1]\n","      if '*' in [orig_amino_acid, comp_amino_acid]:\n","        continue\n","      if right - left == 1:\n","        colors = amino_acid_colors\n","        mutant_color_change = 'same'\n","        if 'X' in [orig_amino_acid, comp_amino_acid]:\n","          mutant_color_change = 'X may differ'\n","        elif colors[orig_amino_acid] != colors[comp_amino_acid]:\n","          mutant_color_change = '{} to {}'.format(colors[orig_amino_acid], colors[comp_amino_acid])\n","        types = amino_acid_types\n","        mutant_change_type = 'conservative'\n","        if 'X' in [orig_amino_acid, comp_amino_acid]:\n","          mutant_change_type = 'X may not be conserved'\n","        elif types[orig_amino_acid] != types[comp_amino_acid]:\n","          mutant_change_type = '{} to {}'.format(types[orig_amino_acid], types[comp_amino_acid])\n","        mutants.append( (left, orig_amino_acid, comp_amino_acid, mutant_color_change, mutant_change_type) )\n","        mutations_count += 1\n","    comp_feat['mutants'] = mutants\n","\n","  # print the features found for strain\n","  print('Mutations found for {} strain: {}'.format(strain, mutations_count))\n","\n","  comparisons = []\n","  for comp in comparison.values():\n","    comparisons.append(comp)\n","    \n","  with open(filepath, 'wb') as file:\n","    pickle.dump(comparisons, file)\n","\n","print('\\nFinding mutations and re-saving files complete')"]},{"attachments":{},"cell_type":"markdown","metadata":{},"source":["### View mutation counts for a strain\n","\n","Here you can select a single strain you want to see found mutations for."]},{"cell_type":"code","execution_count":23,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":172,"status":"ok","timestamp":1683932406997,"user":{"displayName":"Oswald Jones","userId":"18088496464914663925"},"user_tz":420},"id":"EZcaC9CnQR8O","outputId":"6eb02fc3-da0b-465c-fbc1-5f70d147dcec"},"outputs":[{"name":"stdout","output_type":"stream","text":["Amino acid mutation counts found in MW474188.1 from the original NC_045512.2 strain:\n","  surface glycoprotein: 1\n","  leader protein: 0\n","  nsp2: 0\n","  nsp3: 0\n","  nsp4: 0\n","  3C-like proteinase: 0\n","  nsp6: 0\n","  nsp7: 0\n","  nsp8: 0\n","  nsp9: 0\n","  nsp10: 0\n","  helicase: 0\n","  3'-to-5' exonuclease: 0\n","  endoRNAse: 0\n","  2'-O-ribose methyltransferase: 0\n","  ORF1a polyprotein: 0\n","  nsp11: 0\n","  ORF3a protein: 0\n","  envelope protein: 0\n","  membrane glycoprotein: 0\n","  ORF6 protein: 0\n","  ORF7a protein: 0\n","  ORF7b: 0\n","  ORF8 protein: 0\n","  nucleocapsid phosphoprotein: 0\n","  ORF10 protein: 0\n"]}],"source":["# change this to strain you want to see counts for\n","strain = 'MW474188.1'\n","\n","pickles_file = strain + '.pickle'\n","filepath = pickles_path + pickles_file\n","objects = []\n","with (open(filepath, \"rb\")) as openfile:\n","  while True:\n","    try:\n","      objects.append(pickle.load(openfile))\n","    except EOFError:\n","      break\n","comparison_feats = objects[0]\n","\n","print('Amino acid mutation counts found in {} from the original {} strain:'.format(strain, original_strain))\n","for feat in sorted(comparison_feats, key=lambda d: len(d['mutants']), reverse=True):\n","  print('  {}: {}'.format(feat['product'], len(feat['mutants'])))"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyNCNxliFKwkUrxR1wP9/tWi","mount_file_id":"1MQUrXhtZCwsEB15HH3-ZFtUvdHWx0Me_","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.9.1"}},"nbformat":4,"nbformat_minor":0}
